Send message from PRIVLY_APPLICATION to CONTENT_SCRIPT

sammyshj opened the pull request:
Changes done in the PR:
* Does not send message from `CONTENT_SCRIPT` to `CONTENT_SCRIPT` anymore. This is because, the variable `safari.application.activeBrowserWindow.tabs` is not defined when the context of the script is the `CONTENT_SCRIPT`.

sammyshj commented:
@smcgregor Should the messaging to CONTENT_SCRIPT from a CONTENT_SCRIPT be supported? If so, I will update the PR accordingly.

smcgregor commented:
I don't see a reason to prohibit it if an application asks to send it.

sammyshj commented:
@smcgregor Changes made currently:
* The `getContextName()` function has been fixed now.
* From the `PRIVLY_APPLICATION` pages, message cannot be sent to the `CONTENT_SCRIPT`. This is because from the `PRIVLY_APPLICATION` pages context, only the safari variables available are, `safari.extension.baseURI` and `safari.self.tab`, none of which allow to send message to the `CONTENT_SCRIPT`. This can be seen,
![screen shot 2015-07-15 at 5 48 59 am](https://cloud.githubusercontent.com/assets/4390022/8689169/6e13c7cc-2ab5-11e5-951f-75b5fe4ad638.png)

smcgregor commented:
Then how does the iframe resize? Iframe resizing is typically a message from the privly application to the content script.

sammyshj commented:
@smcgregor Can you point me to the code in privly-applications where the iframe resize message is sent? That will help me update this PR.

smcgregor commented:
https://github.com/privly/privly-applications/blob/master/shared/javascripts/host_page_integration.js#L25

sammyshj commented:
@smcgregor Thanks for that. I have updated the code and renamed the PR.
Final changes made in the PR:
* For sending message from `PRIVLY_APPLICATION` to `CONTENT_SCRIPT`, use `parent.postMessage`
* For sending message to `CONTENT_SCRIPT` from other context, use `safari.application.activeBrowserWindow.tabs`
* Fixed the `getContextName()` function to identify `PRIVLY_APPLICATION` appropriately

smcgregor commented:
Does `parent.postMessage` get captured by the content script's scripting environment or the host page's scripting environment? Is it visible to both scripting environments?

sammyshj commented:
The `parent.postMessage` is visible to both the scripting environments.

smcgregor commented:
OK, that is fine since it matches with our threat model for content scripts, but this message interface can't be used for any privileged information.

sammyshj commented:
Sounds good then. I think the PR can be merged now.

smcgregor commented on shared/javascripts/context_messenger.js:
```diff
@@ -425,7 +425,7 @@ if (Privly === undefined) {
     if (window.document.getElementById('is-background-script') !== null) {
       return 'BACKGROUND_SCRIPT';
-    } else if (window.location.href.indexOf(window.location.origin + '/privly-applications') === 0) {
+    } else if (window.location.href.toLowerCase().indexOf(window.location.origin + '/' + window.location.pathname.split('/')[1] + '/privly-applications') === 0) {
```
break this huge statement up.

sammyshj replied:
Made the changes in the latest commit.

smcgregor commented on shared/javascripts/context_messenger.js:
```diff
@@ -444,13 +444,18 @@ if (Privly === undefined) {
-      });
+      if (this.getContextName() === 'PRIVLY_APPLICATION') {
+        // Send message to the parent content script
+        parent.postMessage(payload, '*');
```
Where is parent defined?

sammyshj replied:
parent is defined in the scope of the `PRIVLY_APPLICATION` pages. The parent variable refers to a Window instance.

smcgregor replied:
I am surprised that works on safari. It would be better to be explicit. I think this is short for `window.parent`?

sammyshj replied:
Mentioned window.parent now.

smcgregor commented on shared/javascripts/context_messenger.js:
```diff
@@ -425,10 +425,14 @@ if (Privly === undefined) {
-      return 'CONTENT_SCRIPT';
+      var pathname = window.location.pathname.split('/')[1];
+      var window_location = window.location.origin + '/' + pathname + '/privly-applications';
+      if (window.location.href.toLowerCase().indexOf(window_location) === 0) {
```
Could you change this to `window.location.href.indexOf("safari://") === 0`?

sammyshj replied:
Yes, changed it to safari-extension:// now

sammyshj commented:
@smcgregor Updated the PR.
